### PR TITLE
UI Disabled Inputs

### PR DESCRIPTION
**Pull request by tfamula**
Possibility to make inputs disabled in Forms and (future) Filters.
The need of disabled inputs came up by implementing the [Filter](https://github.com/ILIAS-eLearning/ILIAS/pull/1253#discussion_r228858283)

### CODE REVIEW

File: src/UI/templates/default/Input/tpl.tag_input.html [hunk 1]
```diff
@@ -1,5 +1,5 @@
 <div id="container-{ID}" class="form-control form-control-sm il-input-tag">
-	<input type="text" id="{ID}" value="{VALUE_COMMA_SEPARATED}" class="form-control form-control-sm"/> <input type="hidden" id="template-{ID}" value='{NAME}[]'>
+	<input type="text" id="{ID}" value="{VALUE_COMMA_SEPARATED}"<!-- BEGIN disabled --> {DISABLED}<!-- END disabled --> class="form-control form-control-sm"/> <input type="hidden" id="template-{ID}" value='{NAME}[]'>
```
Comment (tfamula) [inline src/UI/templates/default/Input/tpl.tag_input.html]: Maybe this is not enough here. Technically, the tag-input will be disabled, so that the user cannot modify it. But visually, it still looks enabled because of the < div > over it. The user can also click on it and on the first click, even a pointer comes up.
My idea would be to additionally extend the < div > with a "disabled" class. Adding some css for the color and pointer-event would make the division visually disabled for the user.

**Comment by klees** [comment]
@tfamula Do you possess some mind-reading device? @Amstutz and me talked about this like four hours ago. Thx a lot already, I will have a look into this in the next week. Have a nice weekend later on!

File: src/UI/templates/default/Input/tpl.tag_input.html (hunk 1 above)
Comment (Amstutz) [reply src/UI/templates/default/Input/tpl.tag_input.html]: @tfamula sounds reasonable. @chfsx you have to most know-how about the tax, can you leave a short feedback here?

**Review by Amstutz — CHANGES_REQUESTED** [review]
Hi @tfamula really nice work and awsome timing. As @klees pointed out, where where just writting this on a todo list of ours when this came in. Nice. The interface change lgtm. I think we can discuss this directly at the JF since the change of the interface is really small. I'll attach the according label.

Nice work on the examples. Note that for a merge you need to:

- [ ] Add some test coverage of this new "feature".

**Review by klees — CHANGES_REQUESTED** [review]
LGTM in general, and thx a lot again. 

As far as I understand investigations of @chfsx into the general state of the `disabled`-Attribute, not all browsers will properly display all types of disabled fields, so we might need some additional CSS at some point for some browser.

The changes in this PR currently do not cover `Group`, `Section`, `MultiSelect`, `Numeric` and `Password` (or I am missing something crucial). These are `Input`s (hence get the `disabled`-property by the base class) but currently do not show any meaningful behaviour when the property is set. For `Group` and `Section` users imo would expect, that the contained inputs are disabled, the others should be just disabled like all other fields.

Did you check what happens when you actually process input from disabled fields?

**Comment by matthiaskunkel** [comment]
Jour Fixe, 12 NOV 2018 : We accept this PR for trunk. Thomas will tackle Group and Section inputs, Richard will take care of Required inputs before committing it to trunk.

**Comment by alex40724** [comment]
> Did you check what happens when you actually process input from disabled fields?

Imo we should simply skip disabled inputs in the withRequest/withInput loop. We should not try to pass the values somehow through the request, since we cannot ensure that the values are not manipulated. Consumers must not construct the form differently for the initial presentation and the input processing.

This follows up on the content/data discussion. Imo we need a function that runs through the operations without taking the request data.

So basically a `withInput` without the lines:

```
$value = $post_input->getOr($this->getName(), "");
$clone = $this->withValue($value);
```

`getData` should always deliver all inputs imo.

**Comment by klees** [comment]
@alex40724 Yes, I agree. We should skip the client-server-roundtrip that `ilPropertyFormGUI` performs. I do not expect this roundtrip to happen in the current implementation, though. As @tfamula explained in the JF yesterday, the value set on the disabled input server-side seems to be used in processing of the response, without having been passed to the client in the meantime. This behaviour seems to be correct for me.

**Comment by tfamula** [comment]
Hi @Amstutz and @klees,
we made some modifications to prevent getting the value for disabled inputs from the post-request and instead getting the value from the client side.

We also added the possibility to make `Group`, `Dependant Group` and `Section` disabled. The Inputs, which are located inside these three types of Inputs, will be made disabled technically and visually.
Let me know if you wish additional visual modification for these Inputs (e.g. greying out the whole division). Please consider this would be possible for `Dependant Group` and `Section`, but not for `Group`, because it has no own template and therefore we have no `div` which we can adjust. Since this would not be consistent, I would leave it the way it looks now. 

> As far as I understand investigations of @chfsx into the general state of the disabled-Attribute, not all browsers will properly display all types of disabled fields, so we might need some additional CSS at some point for some browser.

My investigations are different. For all Inputs which we have in the KS currently, the `disabled` attribute is provided by all common browsers from the beginning. (see [here](https://www.w3schools.com/tags/att_disabled.asp) -> Browser Support)

**Comment by chfsx** [comment]
My investigations were for the attribute "readonly" not "disabled": https://www.w3schools.com/tags/att_readonly.asp

**Comment by klees** [comment]
Hi @tfamula, I'm really busy this week, but I promise you to give a feedback till end of the week. Sorry for the delay.

**Comment by tfamula** [comment]
Hi @klees, no problem. I just finished everything that was missing in my eyes and made some small changes here and there.

I would like to say a few words about the Tag Input:
After analysing the html, I found at least five different `input`s, which take part in the Tag Input. I could not find out which of them is the "main input". Trying to give one or more them the `disabled` attribute caused strange behaviour. Moreover, this would not solve our problem anyway: making the input disabled visually. For the request, we do not even need this attribute, because this is handled by the methods `withDisabled` and `isDisabled` technically.
I decided to remove the `disabled` attribute from the chosen `input` and instead extend the class of the `div` with this property. So my solution is to prevent user changes by styling the disabled (pseudo) `input` with the same background-color as it would be with the `disabled` attribute, and making it not usable by mouse, touch and keyboard with the help of css/js.
I hope my proposal is acceptable for you.

File: src/UI/Implementation/Component/Input/Field/Group.php [hunk 2]
```diff
@@ -42,4 +42,15 @@ public function __construct(
 		$this->inputs = $inputs;
 	}
 
+	function withDisabled($is_disabled) {
```
Comment (klees) [inline src/UI/Implementation/Component/Input/Field/Group.php]: Please add `public` here.

File: src/UI/Implementation/Component/Input/Field/Group.php [hunk 3]
```diff
@@ -42,4 +42,15 @@ public function __construct(
...
+		$inputs = [];
+		foreach ($this->inputs as $key => $input)
+		{
+			$inputs[$key] = $input->withDisabled($is_disabled);
+		}
+		$this->inputs = $inputs;
```
Comment (klees) [inline src/UI/Implementation/Component/Input/Field/Group.php]: This violates immutability of the object. Please clone before you assign.

**Review by klees — APPROVED** [review]
LGTM, besides the two minor issues in the comment. Thx a lot again.

@Amstutz: do you want to have another look?

**Review by Amstutz — APPROVED** [review]
Thx a lot for adding the tests. @klees you may merge as soon as your points are taken care of.

**Comment by tfamula** [comment]
Hi @klees,
the immutability violation was solved. We are ready to merge :)


> Richard will take care of Required inputs before committing it to trunk.

I think this is not part of the PR and will be solved by you separately.
